include/texture.hpp

```diff
@@ -33,74 +39,94 @@ struct checker_texture {
-                     bytes_per_pixel);
-
-    if (!data) {
+    uint8_t* _data;
+    int _w, _h;
+    _data =
+        stbi_load(file_name, &_w, &_h, &components_per_pixel, bytes_per_pixel);
+    std::size_t _offset = 0;
+    if (!_data) {
       std::cerr << "ERROR: Could not load texture image file '" << file_name
                 << "'.\n"
                 << stbi_failure_reason() << std::endl;
-      width = height = 0;
+      _w = _h = 1;
+    } else {
+      auto size = bytes_per_pixel * _w * _h;
+      _offset = texture_data.size() / 3;
+      std::copy(_data, _data + size, std::back_inserter(texture_data));
     }
-    // \todo deallocate the image memory in the constructor
+    return image_texture(_w, _h, _offset, _cyclic_frequency);
+  }
 
-    bytes_per_scanline = bytes_per_pixel * width;
+  static sycl::buffer<uint8_t, 2> freeze() {
+    assert(!frozen);
+    frozen = true;
+    return sycl::buffer<uint8_t, 2> { texture_data.data(),
+                                      { texture_data.size() / 3, 3 } };
   }
 
   /// Get the color for the texture at the given place
   /// \todo rename this value() to color() everywhere?
-  color value(const hit_record& rec) const {
+  color value(const hit_record& rec, auto& ctx) const {
     // If texture data is unavailable, return solid cyan
-    if (!data)
-      return { 0, 1, 1 };
-
     // The image is repeated by the repetition factor
-    int i = std::fmod(rec.u * cyclic_frequency, 1) * (width - 1);
+    std::size_t i = std::fmod(rec.u * cyclic_frequency, 1) * (width - 1);
     // The image frame buffer is going downwards, so flip the y axis
-    int j = (1 - std::fmod(rec.v * cyclic_frequency, 1)) * (height - 1);
-
+    std::size_t j = (1 - std::fmod(rec.v * cyclic_frequency, 1)) * (height - 1);
+    std::size_t local_offset = j * width + i;
+    std::size_t pix_idx = local_offset + offset;
     auto scale = 1.f / 255;
-    auto pixel = data + j * bytes_per_scanline + i * bytes_per_pixel;
-    return { pixel[0] * scale, pixel[1] * scale, pixel[2] * scale };
+    auto& texture_data = ctx.texture_data;
+    return { texture_data[pix_idx][0] * scale,
+             texture_data[pix_idx][1] * scale,
+             texture_data[pix_idx][2] * scale };
   }
+
+ private:
+  std::size_t width {};
+  std::size_t height {};
+  static constexpr auto bytes_per_pixel = 3;
+  std::size_t offset;
+
+  static std::vector<uint8_t> texture_data;
+  static bool frozen;
+
+  /// The repetition rate of the image
+  float cyclic_frequency { 1 };
+  image_texture(std::size_t _width, std::size_t _height, std::size_t _offset,
+                float _cyclic_frequency)
+      : width { _width }
+      , height { _height }
+      , offset { _offset }
+      , cyclic_frequency { _cyclic_frequency } {}
 };
 
 using texture_t = std::variant<checker_texture, solid_texture, image_texture>;
 
+std::vector<uint8_t> image_texture::texture_data { 0, 0, 1 };
```

Comment:
why doesn't this start empty ?

Reply:
It is used to have a default texture if the file load doesn't work (which will be solid blue).